Refactor function for kubeconfig

ping035627 opened the pull request:
What type of PR is this?
/kind cleanup

What this PR does / why we need it:
Function 'createKubeConfigFiles' and 'ValidateKubeconfigsForExternalCA' are mostly the same. I think we can refine a new function.

Which issue(s) this PR fixes:

Fixes #

Special notes for your reviewer:

Does this PR introduce a user-facing change?:

/release-note-none
/priority backlog

ping035627 commented:
/release-note-none

neolit123 reviewed (commented):
thank you for the PR @ping035627 

/ok-to-test
@kubernetes/sig-cluster-lifecycle-pr-reviews

ping035627 committed:
Refactor function for kubeconfig

Signed-off-by: PingWang <wang.ping5@zte.com.cn>

update param

Signed-off-by: PingWang <wang.ping5@zte.com.cn>

update function param

Signed-off-by: PingWang <wang.ping5@zte.com.cn>

add comment for const

Signed-off-by: PingWang <wang.ping5@zte.com.cn>

Revert the func param

Signed-off-by: PingWang <wang.ping5@zte.com.cn>

update func define

Signed-off-by: PingWang <wang.ping5@zte.com.cn>

rosti commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:
```diff
@@ -362,25 +338,36 @@ func ValidateKubeconfigsForExternalCA(outDir string, cfg *kubeadmapi.InitConfigu
+}
+
+// createKubeConfigFiles creates all the requested kubeconfig files.
+func handleKubeConfigFiles(fn handleKubeConfig, outDir string, cfg *kubeadmapi.InitConfiguration, kubeConfigFileNames ...string) error {
```
I would rather have the full `func(outDir, filename string, config *clientcmdapi.Config) error` here, than this type.
It's easier to see what the func takes, plus we don't have another type, that's used just here and nowhere else.

neolit123 replied:
@ping035627 
please take a look at @rosti 's comment here. don't define a new type `handleKubeConfig`.

but my preference instead of passing function objects around would be to call break the logic into separate utility functions, even if we have to iterate over list of configs twice. the lists are small...

rosti replied:
@ping035627 ^^

rosti reviewed (approved):
Thanks for the cleanup @ping035627 !
/lgtm

ping035627 commented:
@neolit123 @rosti hi, I changed a different way, we can avoid duplication code, and also see what the func takes. Can you review again? Thanks.

rosti commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:
```diff
@@ -362,25 +343,42 @@ func ValidateKubeconfigsForExternalCA(outDir string, cfg *kubeadmapi.InitConfigu
 		}
 
-		if err = validateKubeConfig(outDir, kubeConfigFileName, kubeconfig); err != nil {
+		switch handleCfgType {
```
I don't like it like this. This narrows down what `handleKubeConfigFiles` can do. Please, revert to the previous version with a function pointer passed to `handleKubeConfigFiles`.

ping035627 replied:
ok, I have revert it.

ping035627 commented:
@neolit123 @rosti hi, can you review this pr again? Thanks.

pytimer commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:
```diff
@@ -362,25 +338,35 @@ func ValidateKubeconfigsForExternalCA(outDir string, cfg *kubeadmapi.InitConfigu
 
-		kubeconfig, err := buildKubeConfigFromSpec(spec, cfg.ClusterName)
+		// builds the KubeConfig object
+		config, err := buildKubeConfigFromSpec(spec, cfg.ClusterName)
```
I think here can still using the previous variable `kubeconfig`, no need to change it to `config`. What do you think?

ping035627 replied:
This variable name is different in two functions that are almost repetitive. I just use one of the variable names, which is not modifiable.

pytimer replied:
I see that the two functions are almost repetitive, just think that you maybe using the logic of `ValidateKubeconfigsForExternalCA` function, but this does not affect.

rosti replied:
I agree with @pytimer . Let's use the `kubeconfig` name here.
We have both `cfg` and `config` names which point to different things and this is not good.

rosti commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:
```diff
@@ -39,6 +39,8 @@ import (
 	kubeconfigutil "k8s.io/kubernetes/cmd/kubeadm/app/util/kubeconfig"
 )
 
+type handleKubeConfig func(outDir, filename string, config *clientcmdapi.Config) error
```
Let's not define this type.

pytimer replied:
If we remove this type, the invoke this type of  functions parameters maybe looks too long.

rosti replied:
Yes, the declaration will be longer, but it will also be more descriptive. Now you have to go and look at what this type is to deduce it's a function. Also, we are polluting the code a bit if we define a type to used only at a single spot and nowhere else.

pytimer replied:
Yes, you right. This way may have less pollution to the code.

ping035627 commented:
@neolit123 @rosti hi, I have modified it according to your opinion. Can you review again? Thanks.

ping035627 commented:
/retest

rosti reviewed (approved):
Perfect! Thank you @ping035627 !
/lgtm

neolit123 commented:
i'd still prefer the refactor that i tried to explain here:
https://github.com/kubernetes/kubernetes/pull/74483#discussion_r261478128

ping035627 commented:
@neolit123 I think it's not easy to break the logic into separate utility functions  in this place, it's in for loop. Do you have good idea? Thanks.

timothysc reviewed (approved):
/approve

rosti commented:
@ping035627 can you please rebase this PR and ping me for lgtm?

ping035627 commented:
@rosti hi, I rebased the code, and I found that it was no longer necessary to extract functions from the rebased code. I think we should close this PR. Thanks.

neolit123 commented:
@ping035627 
ok, thanks for confirming.
/close
